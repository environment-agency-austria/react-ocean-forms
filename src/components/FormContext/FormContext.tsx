/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';
import { IFormContext } from './FormContext.types';

/**
 * Context for the communication between the form
 * and its fields.
 */
export const FormContext = React.createContext<IFormContext | undefined>(undefined);
