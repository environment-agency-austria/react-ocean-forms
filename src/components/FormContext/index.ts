export * from './FormContext';
export * from './FormContext.types';
export * from './FormContext.hooks';
