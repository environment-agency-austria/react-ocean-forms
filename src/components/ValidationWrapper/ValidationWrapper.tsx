import React from 'react';

import { useFullName, useValidation } from '../../hooks';
import { IValidationProp, IValidationWrapperProps } from './ValidationWrapper.types';

/**
 * Component that handles validation of the
 * wrapped component.
 */
export const ValidationWrapper: React.FC<IValidationWrapperProps> = ({ name, validators, asyncValidators, asyncValidationWait, render }): JSX.Element => {
  const fullName = useFullName(name);

  const [
    validationState,
    validate,
    reset,
    update,
  ] = useValidation(
    fullName,
    validators,
    asyncValidators,
    asyncValidationWait
  );

  const validationProp: IValidationProp = {
    ...validationState,
    reset,
    validate,
    update,
  };

  return render(
    fullName,
    validationProp,
  );
};
