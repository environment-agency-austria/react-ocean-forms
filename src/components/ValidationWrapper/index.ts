export * from './ValidationWrapper';
export * from './ValidationWrapper.types';
