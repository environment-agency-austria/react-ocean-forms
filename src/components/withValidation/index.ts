export * from './withValidation';
export * from './withValidation.types';
