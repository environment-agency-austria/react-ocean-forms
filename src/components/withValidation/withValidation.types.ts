import { TAsyncValidator, TFieldErrors, TValidator } from '../../validators';
import { TFieldValue } from '../Field';
import { IFormContextProps } from '../withForm';

/**
 * Properties of a component that is wrapped
 * by withValidation
 */
export interface IValidatedComponentProps extends IFormContextProps {
  /**
   * Field name
   */
  name: string;
  /**
   * Synchronous validators
   */
  validators?: TValidator[];
  /**
   * Asynchronous validators
   */
  asyncValidators?: TAsyncValidator[];
  /**
   * Wait time in ms that should pass after
   * the last user input before the async
   * validators will be triggered
   */
  asyncValidationWait?: number;
}

/**
 * Interface with properties describing the current
 * validation state and offering interfaces for
 * various validation tasks
 */
export interface IValidationProp extends IValidationState {
  /**
   * Triggers the validation of the field
   * @param value Field value
   * @param args Validation args @see IValidationArgs
   */
  validate(value: TFieldValue, args?: Partial<IValidationArgs>): Promise<IValidationState>;
  /**
   * Resets the validation state
   */
  reset(): void;
  /**
   * Forces a new validation state on this Field
   * @param state New validation state
   */
  update(state: Partial<IValidationState>): void;
}

/**
 * Base interface for consumers of withValidation props
 */
export interface IBaseValidationProps {
  /**
   * Full Name of the component
   * (context.fieldPrefix + '.' + fieldName)
   */
  fullName: string;
  /**
   * Validation properties, describes the current
   * validation state of the component
   */
  validation: IValidationProp;
}

/**
 * Interface for consumers of withValidation props
 * Note: this interface is including the form context
 * props, which are automatically mapped by the
 * withValidation hoc
 */
export interface IValidationProps extends IBaseValidationProps, IFormContextProps { }

/**
 * Internal state of the validated component
 */
export interface IValidationState {
  /**
   * True, if the field is currently validating
   * (asynchronous validation running in background)
   */
  isValidating: boolean;
  /**
   * True, if all validators report a valid state
   */
  valid: boolean;
  /**
   * Contains any errors if available
   */
  error: TFieldErrors;
}

/**
 * Arguments for the validate method
 */
export interface IValidationArgs {
  /**
   * True, if the async validators should
   * be triggered as well, otherwise only
   * the sync validators are triggered
   *
   * Default: true
   */
  checkAsync: boolean;
  /**
   * True, if the async validators should
   * be triggered without any delay
   *
   * Default: false
   */
  immediateAsync: boolean;
}
