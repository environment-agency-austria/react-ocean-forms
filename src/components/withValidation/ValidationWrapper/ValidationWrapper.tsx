import * as React from 'react';

import { parseValidationError } from '../../../utils';
import { isIFieldErrorObject } from '../../../validators';
import { TBasicFieldValue } from '../../withField';
import { withForm } from '../../withForm';
import { IValidationArgs, IValidationState, IValidationWrapperProps } from '../withValidation.types';

/**
 * Component that handles validation of the
 * wrapped component.
 */
export class BaseValidationWrapper extends React.Component<IValidationWrapperProps, IValidationState> {
  public static displayName: string = 'ValidationWrapper';

  private unmounted: boolean = false;

  private asyncTimeout?: number;

  private get fullName(): string {
    const { name, context } = this.props;

    return context.fieldPrefix !== null
      ? context.fieldPrefix.concat('.', name)
      : name;
  }

  constructor(props: IValidationWrapperProps) {
    super(props);

    this.state = {
      valid: true,
      error: null,
      isValidating: false,
    };
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    this.clearValidationTimeout();
    this.unmounted = true;
  }

  /**
   * Returns the current async timeout
   */
  public getAsyncTimeout(): number | undefined {
    return this.asyncTimeout;
  }

  /**
   * Clears the validation timeout if currently
   * running
   */
  private clearValidationTimeout(): void {
    if (this.asyncTimeout !== undefined) {
      clearTimeout(this.asyncTimeout);
      this.asyncTimeout = undefined;
    }
  }

  /**
   * Resets the validation state to the default
   */
  private reset = (): void => {
    this.clearValidationTimeout();

    this.updateAndNotify({
      valid: true,
      error: null,
      isValidating: false,
    });
  }

  /**
   * Updates the validation state
   * @param state New state
   */
  private updateValidationState = (state: Partial<IValidationState>): void => {
    const oldState = this.state;
    const newState = {
      ...oldState,
      ...state,
    };

    this.updateAndNotify(newState);
  }

  /**
   * Triggers the validation of a field.
   * @param value Value of the field
   * @param options Validation params @see checkAsync @see immediateAsync
   * @param checkAsync True if the async validators should be triggered as well
   * @param immediateAsync True if the async validators should fire immediately
   */
  // tslint:disable-next-line:max-func-body-length
  private validate = async (
    value: TBasicFieldValue,
    {
      checkAsync = true,
      immediateAsync = false,
    }: Partial<IValidationArgs> = {},
  ): Promise<IValidationState> => {
    const {
      validators,
      asyncValidators,
      context: formContext,
    } = this.props;

    const validationState: IValidationState = {
      valid: true,
      error: null,
      isValidating: false,
    };

    // Clear the old timeout so we only run the
    // async validators after the waiting period
    // when the value didn't change in the meantime
    this.clearValidationTimeout();

    // No validators - nothing to do here
    if (!Array.isArray(validators) && !Array.isArray(asyncValidators)) {
      this.setState(validationState);

      return validationState;
    }

    // Synchronous validators
    if (Array.isArray(validators)) {
      validationState.valid = validators.every((validator) => {
        const result = validator(value, formContext);
        const parsedResult = parseValidationError(this.fullName, result);

        if (isIFieldErrorObject(parsedResult)) {
          validationState.error = parsedResult;

          return false;
        }

        return true;
      });
    }

    // Ignore async validation if sync validation is already false
    if (validationState.valid === false) {
      this.updateAndNotify(validationState);

      return validationState;
    }

    // Only run async validation if needed
    if (!checkAsync || !Array.isArray(asyncValidators)) {
      this.updateAndNotify(validationState);

      return validationState;
    }

    // Asynchronous validators
    const performAsyncValidation = async (): Promise<IValidationState> => {
      const validatorFunctions = asyncValidators.map(async validator => validator(
        value,
        formContext,
      ));

      const errors = await Promise.all(validatorFunctions);
      const parsedErrors = errors.map(error => parseValidationError(this.fullName, error));

      validationState.error = parsedErrors.filter(isIFieldErrorObject);
      validationState.valid = validationState.error.length === 0;

      if (validationState.error.length === 0) { validationState.error = null; }

      validationState.isValidating = false;
      this.asyncTimeout = undefined;

      this.updateAndNotify(validationState);

      return validationState;
    };

    validationState.isValidating = true;

    if (immediateAsync === true) {
      this.updateAndNotify(validationState);

      return performAsyncValidation();
    }

    // Get the correct wait setting
    const { asyncValidationWait: propAsyncValidationWait } = this.props;
    const asyncValidationWait = propAsyncValidationWait === null
      ? formContext.asyncValidationWait
      : propAsyncValidationWait;

    this.asyncTimeout = setTimeout(performAsyncValidation, asyncValidationWait);

    this.updateAndNotify(validationState);

    return validationState;
  }

  /**
   * Updates the validation state and notifies the form
   * @param newState New validation state
   */
  private updateAndNotify(newState: IValidationState): void {
    const { context } = this.props;

    // Don't do anything if the component has already been
    // unmounted. This can happen when the validated Field
    // is already removed while there are async validators
    // running in the background.
    if (this.unmounted) { return; }

    this.setState(newState);
    context.notifyFieldEvent(this.fullName, 'validation', newState);
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      isValidating,
      valid,
      error,
    } = this.state;

    const { context } = this.props;

    const validation = {
      isValidating,
      valid,
      error,

      validate: this.validate,
      reset: this.reset,
      update: this.updateValidationState,
    };

    return this.props.render(
      this.fullName,
      validation,
      context,
    );
  }
}

export const ValidationWrapper = withForm(BaseValidationWrapper);
