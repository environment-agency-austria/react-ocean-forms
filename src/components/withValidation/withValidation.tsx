/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { getDisplayName, PropsOf, Subtract } from '../../utils';
import { IFormContext } from '../FormContext';
import { ValidationWrapper } from './ValidationWrapper';
import { IValidatedComponentProps, IValidationProp, IValidationProps } from './withValidation.types';

type WrappedValidatedComponentProps<TComp> =
  Subtract<JSX.LibraryManagedAttributes<TComp, PropsOf<TComp>>, IValidationProps> & IValidatedComponentProps;

/**
 * Higher order component for validation
 */
export const withValidation = <TComp extends React.ComponentType<TProps>, TProps extends IValidationProps = PropsOf<TComp>>
(component: TComp): React.ComponentType<WrappedValidatedComponentProps<TComp>> => {

  // tslint:disable-next-line:naming-convention
  const CastedComponent = component as React.ComponentType<TProps>;

  type IWrappedProps = WrappedValidatedComponentProps<TComp>;

  const validatedComponent: React.SFC<IWrappedProps> = (props: IWrappedProps): JSX.Element => {
    const renderComponent = (fullName: string, validation: IValidationProp, context: IFormContext): JSX.Element => {
      // @ts-ignore
      return <CastedComponent context={context} fullName={fullName} validation={validation} {...props} />;
    };

    return (
      <ValidationWrapper
        {...props}
        render={renderComponent}
      />
    );
  };
  validatedComponent.displayName = `ValidatedComponent(${getDisplayName(component)})`;

  return validatedComponent;
};
