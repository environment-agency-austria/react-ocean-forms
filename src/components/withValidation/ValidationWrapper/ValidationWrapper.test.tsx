import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { createMockFormContext } from '../../../test-utils/enzymeFormContext';
import { validators as defaultValidators } from '../../../validators';
import { IFormContext } from '../../FormContext';
import { IValidationProp, IValidationState, IValidationWrapperProps } from '../withValidation.types';
import { BaseValidationWrapper } from './ValidationWrapper';

describe('withValidation', () => {
  const fieldName = 'unitField';

  interface ISetupArgs {
    props?: Partial<IValidationWrapperProps>;
    contextOverrides?: Partial<IFormContext>;
  }

  interface ISetupResult {
    formContext: IFormContext;
    validation: IValidationProp;
    wrapper: ShallowWrapper;
    fullName: string;
  }

  const setup = ({
    props,
    contextOverrides,
  }: Partial<ISetupArgs> = {}): ISetupResult => {
    const formContext: IFormContext = {
      ...createMockFormContext(),
      ...contextOverrides,
    };

    let fullName: string;
    let validation: IValidationProp;

    const renderCallback = (cFullName: string, cValidation: IValidationProp, cContext: IFormContext): JSX.Element => {
      fullName = cFullName;
      validation = cValidation;

      return (
        <div id="test-component" />
      );
    };

    const wrapper = shallow((
      <BaseValidationWrapper
        name={fieldName}
        context={formContext}
        render={renderCallback}
        {...props}
      />
    ));

    return {
      formContext,
      // @ts-ignore
      validation,
      wrapper,
      // @ts-ignore
      fullName,
    };
  };

  it('should render without error', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should return a valid state without validators', async () => {
    const { validation } = setup();
    await expect(validation.validate('foo')).resolves.toMatchObject({
      error: null,
      isValidating: false,
      valid: true,
    });
  });

  const mockValue = 'foobar';
  const checkNotifyCalled = (formContext: IFormContext, state: IValidationState): void => {
    expect(formContext.notifyFieldEvent).toHaveBeenLastCalledWith(
      fieldName,
      'validation',
      state,
    );
  };

  const getAsyncTimeout = (wrapper: ShallowWrapper): number | undefined => (wrapper.instance() as BaseValidationWrapper).getAsyncTimeout();

  describe('sync validation', () => {
    it('should call the sync validators and return a validation state', async () => {
      const validator = jest.fn().mockReturnValue(undefined);

      const { validation, formContext } = setup({ props: {
        validators: [validator],
      }});

      const state = await validation.validate(mockValue);
      expect(state).toMatchObject({
        isValidating: false,
        valid: true,
        error: null,
      });

      checkNotifyCalled(formContext, state);
    });

    it('should stop at the first invalid validator', async () => {
      const errorId = 'mockError';
      const validators = [
        jest.fn().mockReturnValue(undefined),
        jest.fn().mockReturnValue(errorId),
        jest.fn().mockReturnValue(undefined),
      ];

      const { validation } = setup({ props: {
        validators,
      }});

      await validation.validate(mockValue);

      expect(validators[0]).toHaveBeenCalledTimes(1);
      expect(validators[1]).toHaveBeenCalledTimes(1);
      expect(validators[2]).not.toHaveBeenCalled();
    });

    it('should not call the async validator if the sync validators are invalid', async () => {
      const validator = jest.fn().mockReturnValue('error');
      const asyncValidator = jest.fn().mockResolvedValue(undefined);

      const { validation } = setup({ props: {
        validators: [validator],
        asyncValidators: [asyncValidator],
      }});

      await validation.validate(mockValue);

      expect(asyncValidator).not.toHaveBeenCalled();
    });

    describe('required validator', () => {
      const createValidator = (defaultValue: unknown): Function => {
        const customRequiredValidator = (): undefined => undefined;
        customRequiredValidator.isDefaultValidator = defaultValue;

        return customRequiredValidator;
      };

      const cases = [
        [
          'should set validation.isRequired to false if there is no required validator present',
          undefined,
          false,
        ],
        [
          'should set validation.isRequired to true if the default validator is present',
          [ defaultValidators.required ],
          true,
        ],
        [
          'should set validation.isRequired to true if a validator with isDefaultValidator=true is present',
          [ createValidator(true) ],
          true,
        ],
        [
          'should set validation.isRequired to false if a validator with isDefaultValidator=false is present',
          [ createValidator(false) ],
          false,
        ],
        [
          'should set validation.isRequired to false if a validator with isDefaultValidator="foobar" is present',
          [ createValidator('foobar') ],
          false,
        ],
      ];

      it.each(cases)('%s', (name, validators, expectedIsRequiredState) => {
        const { validation } = setup({ props: {
          validators,
        }});

        expect(validation.isRequired).toEqual(expectedIsRequiredState);
      });
    });
  });

  describe('async validation', () => {
    beforeEach(jest.useFakeTimers);
    afterEach(() => {
      // We must clear and reset the fake timers in order
      // to be able to check if setTimeout has been called
      // for every test. Otherwise it keeps this information
      // through multiple tests.
      jest.clearAllTimers();
      jest.useRealTimers();
    });

    it('should ignore async validators if checkAsync is false', async () => {
      const asyncValidator = jest.fn().mockResolvedValue(undefined);

      const { validation, formContext } = setup({ props: {
        asyncValidators: [asyncValidator],
      }});

      const state = await validation.validate(mockValue, { checkAsync: false });
      expect(state).toMatchObject({
        isValidating: false,
        valid: true,
        error: null,
      });
      expect(asyncValidator).not.toHaveBeenCalled();

      checkNotifyCalled(formContext, state);
    });

    it('should immediately run the validators if immediateAsync is true', async () => {
      const errorId = 'mockError';
      const asyncValidator = jest.fn().mockResolvedValue(errorId);

      const { validation, formContext } = setup({ props: {
        asyncValidators: [asyncValidator],
      }});

      const state = await validation.validate(mockValue, { immediateAsync: true });
      expect(state).toMatchObject({
        isValidating: false,
        valid: false,
        error: [{
          message_id: errorId,
          params: {},
        }],
      });
      expect(asyncValidator).toHaveBeenCalledTimes(1);

      checkNotifyCalled(formContext, state);
    });

    it('should wait for the default amount until triggering the async validators', async (done) => {
      const errorId = 'mockError';
      const asyncValidator = jest.fn().mockResolvedValue(errorId);

      const { wrapper, validation, formContext } = setup({ props: {
        asyncValidators: [asyncValidator],
      }});

      const state = await validation.validate(mockValue);

      expect(state).toMatchObject({
        isValidating: true,
        valid: true,
        error: null,
      });
      expect(asyncValidator).not.toHaveBeenCalled();
      expect(getAsyncTimeout(wrapper)).toBeGreaterThan(0);
      checkNotifyCalled(formContext, state);

      jest.runAllTimers();

      process.nextTick(() => {
        expect(state).toMatchObject({
          isValidating: false,
          valid: false,
          error: [{
            message_id: errorId,
            params: {},
          }],
        });
        expect(asyncValidator).toHaveBeenCalledTimes(1);
        checkNotifyCalled(formContext, state);
        done();
      });
    });

    it('should clear any existing timeout if validate is called again', async () => {
      const errorId = 'mockError';
      const asyncValidator = jest.fn().mockResolvedValue(errorId);

      const { wrapper, validation } = setup({ props: {
        asyncValidators: [asyncValidator],
      }});

      expect(getAsyncTimeout(wrapper)).toBeUndefined();

      const state1 = await validation.validate(mockValue);
      const timeout1 = getAsyncTimeout(wrapper);

      const state2 = await validation.validate(mockValue);
      const timeout2 = getAsyncTimeout(wrapper);

      expect(state1).toMatchObject({
        isValidating: true,
      });
      expect(state2).toMatchObject({
        isValidating: true,
      });

      expect(timeout1).not.toEqual(timeout2);
      jest.runAllTimers();
    });
  });

  describe('form context callbacks', () => {
    it('should update the validation state if called for', () => {
      const { formContext, validation } = setup();

      const mockError = {
        message_id: 'dummy',
        params: {},
      };
      validation.update({
        valid: false,
        error: mockError,
      });
      checkNotifyCalled(
        formContext,
        {
          valid: false,
          error: mockError,
          isValidating: false,
          isRequired: false,
        },
      );
    });

    it('should correctly reset the validation state', () => {
      const { wrapper, formContext, validation } = setup();

      validation.reset();
      checkNotifyCalled(
        formContext,
        {
          valid: true,
          error: null,
          isValidating: false,
          isRequired: false,
        },
      );
      expect(getAsyncTimeout(wrapper)).toBeUndefined();

      // Edge case where we check if the timeout has
      // been cleared if a validation is in progress
      // when we call reset
      void validation.validate(mockValue);
      validation.reset();
      checkNotifyCalled(
        formContext,
        {
          valid: true,
          error: null,
          isValidating: false,
          isRequired: false,
        },
      );
      expect(getAsyncTimeout(wrapper)).toBeUndefined();
    });
  });

  it('should clear any timeouts on unmount', () => {
    const { wrapper, validation } = setup({
      props: {
        asyncValidators: [ jest.fn().mockResolvedValue(undefined) ],
      },
    });

    void validation.validate(mockValue);
    expect(getAsyncTimeout(wrapper)).not.toBeUndefined();

    const oldInstance = wrapper.instance() as BaseValidationWrapper;
    wrapper.unmount();

    const asyncTimeout = oldInstance.getAsyncTimeout();
    expect(asyncTimeout).toBe(undefined);
  });

  it('should adapt the fieldPrefix to the fullName', () => {
    const { fullName } = setup({
      contextOverrides: {
        fieldPrefix: 'unit',
      },
    });

    expect(fullName).toBe(`unit.${fieldName}`);
  });

  it('should not update its state after unmounting', () => {
    const { wrapper, formContext } = setup();

    const oldInstance = wrapper.instance() as BaseValidationWrapper;
    wrapper.unmount();

    expect(() => {
      // @ts-ignore
      oldInstance.updateAndNotify({ foo: 'bar' });
    }).not.toThrowError();
    expect(formContext.notifyFieldEvent).not.toHaveBeenCalled();
  });
});
