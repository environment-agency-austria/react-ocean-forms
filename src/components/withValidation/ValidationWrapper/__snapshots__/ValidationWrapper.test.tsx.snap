// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withValidation should render without error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ValidationWrapper
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction],
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    name="unitField"
    render={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "id": "test-component",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "id": "test-component",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
